src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,320 @@
+import lanchon.multidexlib2.*
+import org.jf.dexlib2.iface.DexFile
+import org.jf.dexlib2.writer.io.MemoryDataStore
+import java.io.File
+
+
+/**
+ * An [Apk] file.
+ */
+sealed class Apk private constructor(module: ApkModule) {
+    companion object {
+        const val manifest = "AndroidManifest.xml"
+    }
+
+    private val archive = Archive(module)
+
+    internal var logger: Logger? = null
+
+    /**
+     * The metadata of the [Apk].
+     */
+    val packageMetadata = PackageMetadata(module.androidManifestBlock)
+
+
+    /**
+     * Update the [PackageBlock] name to match the manifest.
+     */
+    protected open fun updatePackageBlock() {
+        resources.packageBlock!!.name = archive.readManifest().packageName
+    }
+
+    /**
+     * Write the [Apk] to a file.
+     *
+     * @param file The target file.
+     */
+    internal open fun write(file: File) {
+        archive.cleanup(logger)
+
+        updatePackageBlock()
+
+        archive.save(file)
+    }
+
+    inner class ResourceContainer internal constructor(packageName: String, internal val tableBlock: TableBlock?) {
+        internal val packageBlock: PackageBlock? =
+            tableBlock?.packageArray?.let { array ->
+                // Full apks with embedded dynamic feature modules may have multiple package blocks.
+                // The name of the main package block is always the same as the package name in the manifest.
+                if (array.childes.size == 1) array[0] else array.iterator()?.asSequence()
+                    ?.single { it.name == packageName }
+            }
+
+        internal lateinit var resourceTable: ApkBundle.ResourceTable
+
+        private fun expectPackageBlock() = packageBlock ?: throw ApkException.MissingResourceTable
+
+        internal fun getOrCreateTableString(value: String) =
+            tableBlock?.stringPool?.getOrCreate(value) ?: throw ApkException.MissingResourceTable
+
+        private fun Entry.setTo(value: Resource) {
+            val savedRef = specReference
+            ensureComplex(value.complex)
+            if (savedRef != 0) {
+                // Preserve the entry name by restoring the previous spec block reference (if present).
+                specReference = savedRef
+            }
+
+            logger?.trace("Updated @$typeName/$name")
+
+            value.write(this, this@ResourceContainer)
+            resourceTable.registerChanged(this)
+        }
+
+        private fun getEntry(type: String, name: String, qualifiers: String?): Entry? {
+            val resourceId = try {
+                resourceTable.resolve("@$type/$name")
+            } catch (_: ApkException.InvalidReference) {
+                return null
+            }
+
+            val config = ResConfig.parse(qualifiers)
+            return tableBlock?.resolveReference(resourceId)?.singleOrNull { it.resConfig == config }
+        }
+
+        private fun createHandle(resPath: String): ResourceFile.Handle {
```

Comment:
This class implements a lot of methods, that are not suitable in this class but instead in a seperate module that wraps around arsclib. Please move as much as arsclib related code to its own module.